File: frontend/src/api/bunker_transport.ts
import apiClient from './client'

export interface BunkerTransport {
  id: number
  date: string
  time?: string
  truck_id: number
  driver_id: number
  origin: string
  destination: string
  tonnage_kg: number
  billed_tonnage_kg?: number
  cost_per_kg: number
  receipt_no: string
  bill_of_lading_url?: string
  is_dead_freight: boolean
  is_paid: boolean
  payment_date?: string
  payment_receipt_url?: string
  notes?: string
  truck?: { id: number; plate_number: string }
  driver?: { id: number; full_name: string }
  created_at: string
}

export interface BunkerTransportCreate {
  date: string
  time?: string
  truck_id: number
  driver_id: number
  origin: string
  destination: string
  tonnage_kg: number
  billed_tonnage_kg?: number
  cost_per_kg: number
  receipt_no: string
  is_dead_freight?: boolean
  notes?: string
}

export interface BunkerTransportPayment {
  payment_date: string
  notes?: string
}

export interface PaginatedBunkerTransports {
  items: BunkerTransport[]
  total: number
  page: number
  size: number
}

export const getBunkerTransports = (params?: Record<string, unknown>) =>
  apiClient.get<PaginatedBunkerTransports>('/bunker-transport', { params }).then(r => r.data)

export const getBunkerTransport = (id: number) =>
  apiClient.get<BunkerTransport>(`/bunker-transport/${id}`).then(r => r.data)

export const createBunkerTransport = (data: BunkerTransportCreate) =>
  apiClient.post<BunkerTransport>('/bunker-transport', data).then(r => r.data)

export const updateBunkerTransport = (id: number, data: Partial<BunkerTransportCreate>) =>
  apiClient.put<BunkerTransport>(`/bunker-transport/${id}`, data).then(r => r.data)

export const payBunkerTransport = (id: number, data: BunkerTransportPayment) =>
  apiClient.post<BunkerTransport>(`/bunker-transport/${id}/pay`, data).then(r => r.data)

export const deleteBunkerTransport = (id: number) =>
  apiClient.delete(`/bunker-transport/${id}`)

export const uploadBillOfLading = (id: number, file: File) => {
  const fd = new FormData()
  fd.append('file', file)
  return apiClient.post<BunkerTransport>(`/bunker-transport/${id}/bill-of-lading`, fd).then(r => r.data)
}

export const uploadPaymentReceipt = (id: number, file: File) => {
  const fd = new FormData()
  fd.append('file', file)
  return apiClient.post<BunkerTransport>(`/bunker-transport/${id}/payment-receipt`, fd).then(r => r.data)
}

export const importBunkerTransports = (rows: Record<string, unknown>[]) =>
  apiClient.post<{ success: number; errors: string[] }>('/bunker-transport/import', { rows }).then(r => r.data)

File: backend/app/schemas/bunker_transport.py
import uuid
from datetime import date, datetime
from datetime import time as Time

from pydantic import BaseModel


class BunkerTransportBase(BaseModel):
    date_jalali: str
    date_gregorian: date
    time: Time | None = None
    truck_id: uuid.UUID
    driver_id: uuid.UUID
    receipt_no: str
    tonnage_kg: int
    origin_id: uuid.UUID
    cost_per_kg: int
    notes: str | None = None


class BunkerTransportCreate(BunkerTransportBase):
    pass


class BunkerTransportUpdate(BaseModel):
    date_jalali: str | None = None
    date_gregorian: date | None = None
    time: Time | None = None
    truck_id: uuid.UUID | None = None
    driver_id: uuid.UUID | None = None
    receipt_no: str | None = None
    tonnage_kg: int | None = None
    origin_id: uuid.UUID | None = None
    cost_per_kg: int | None = None
    notes: str | None = None


class BunkerTransportPaymentUpdate(BaseModel):
    payment_date_jalali: str
    payment_date_gregorian: date


class TruckInfo(BaseModel):
    id: uuid.UUID
    plate_number: str

    model_config = {"from_attributes": True}


class DriverInfo(BaseModel):
    id: uuid.UUID
    full_name: str

    model_config = {"from_attributes": True}


class SiteInfo(BaseModel):
    id: uuid.UUID
    code: str
    name_fa: str
    name_en: str

    model_config = {"from_attributes": True}


class BunkerTransportResponse(BunkerTransportBase):
    id: uuid.UUID
    is_dead_freight: bool
    billed_tonnage_kg: int
    bill_of_lading_image: str | None
    status: str
    payment_date_jalali: str | None
    payment_date_gregorian: date | None
    payment_receipt_image: str | None
    is_paid: bool
    created_at: datetime
    updated_at: datetime
    created_by: uuid.UUID | None
    truck: TruckInfo | None
    driver: DriverInfo | None
    origin: SiteInfo | None

    model_config = {"from_attributes": True}


class BunkerTransportListResponse(BaseModel):
    items: list[BunkerTransportResponse]
    total: int
    page: int
    size: int
